### PR TITLE
Isssue/211 geb tests

### PR DESCRIPTION
**Pull request by thiefik**
Added tests in GEB. They are working on Travis CI, also they're starting up 'Swagger UI' before them.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -117,6 +117,14 @@ dependencies {
         exclude group: 'org.mortbay.jetty', module: 'servlet-api'
     }
     testCompile "com.ofg:micro-infra-spring-test:$microInfraSpringVersion"
+
+    testCompile "org.gebish:geb-core:0.10.0"
+    testCompile "org.gebish:geb-spock:0.10.0"
+    testCompile "org.spockframework:spock-spring"
+    testCompile "org.seleniumhq.selenium:selenium-firefox-driver:2.43.1"
+    testCompile "org.seleniumhq.selenium:selenium-support:2.43.1"
+    testCompile "org.seleniumhq.selenium:selenium-htmlunit-driver:2.43.1"
```
Comment (szpak) [inline build.gradle]: I would move repeatable versions to gradle.properties (like `microInfraSpringVersion` already).

File: src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.test.context.web.WebAppConfiguration
+
+@ContextConfiguration(loader = SpringApplicationContextLoader.class, classes = Application)
+@WebAppConfiguration
+@IntegrationTest("spring.profiles.active:dev,stubrunner.skip-local-repo:true")
+class AcceptanceSwaggerUISpec extends GebSpec {
+
+    def "SwaggerUI home page should be visible"() {
+        when:
+        to SwaggerUIHomePage
```
Comment (szpak) [inline src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy]: I would propose more compact Spock specification formatting already used n other places: https://solidsoft.wordpress.com/2014/05/20/more-compact-automatic-spock-specification-formatting-in-idea-13/

File: src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package com.ofg.geb
```
Comment (szpak) [inline src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy]: I would move it to `com.ofg.twitter.geb` package. For now it is specific to that application. Maybe in the future we would like to move common things to acceptance-test submodule in micro-infra-spring.

File: src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import com.ofg.twitter.Application
+import geb.spock.GebSpec
+import org.springframework.boot.test.IntegrationTest
+import org.springframework.boot.test.SpringApplicationContextLoader
+import org.springframework.test.context.ContextConfiguration
+import org.springframework.test.context.web.WebAppConfiguration
+
+@ContextConfiguration(loader = SpringApplicationContextLoader.class, classes = Application)
+@WebAppConfiguration
+@IntegrationTest("spring.profiles.active:dev,stubrunner.skip-local-repo:true")
```
Comment (szpak) [inline src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy]: Nice idea with using Spring mechanism to setup the application.

File: src/test/groovy/com/ofg/geb/pages/SwaggerUIHomePage.groovy [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package com.ofg.geb.pages
+
+import geb.Page
+
+class SwaggerUIHomePage extends Page {
+    static url = "http://localhost:8095/swagger"
```
Comment (szpak) [inline src/test/groovy/com/ofg/geb/pages/SwaggerUIHomePage.groovy]: As a next step host and port could be parametrized.

File: src/test/groovy/com/ofg/geb/MicroserviceControllerUISpec.groovy [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+
+    }
+
+    def "Try it out and Check respond body and code"() {
+        when:
+        microserviceGetTryButton.click()
+        waitFor { microserviceGetResponseCode.displayed }
+        then:
+        assert microserviceGetResponseCode.displayed
```
Comment (szpak) [inline src/test/groovy/com/ofg/geb/MicroserviceControllerUISpec.groovy]: `assert` is not redundant here

File: src/test/groovy/com/ofg/geb/MicroserviceControllerUISpec.groovy [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        waitFor { microserviceGetTryButton.displayed }
+        microserviceGetTryButton.displayed
+
+
+    }
+
+    def "Try it out and Check respond body and code"() {
+        when:
+        microserviceGetTryButton.click()
+        waitFor { microserviceGetResponseCode.displayed }
```
Comment (szpak) [inline src/test/groovy/com/ofg/geb/MicroserviceControllerUISpec.groovy]: Maybe move it to `then` section?

File: src/test/groovy/com/ofg/geb/MicroserviceControllerUISpec.groovy [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        then:
+        assert microserviceGetResponseCode.displayed
+        assert microserviceGetResponseCode.text() == '200'
+        def inputFile = new File("src/main/resources/microservice.json")
```
Comment (szpak) [inline src/test/groovy/com/ofg/geb/MicroserviceControllerUISpec.groovy]: Maybe reading from a classpath to be not working path dependent?

File: src/test/groovy/com/ofg/geb/HealthEndpointUISpec.groovy [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    def "Check visibility of operations and their paths for 'health-mvc-endpoint' #httpOperation"() {
+
+        expect:
+        getTextfromHealthOperation(httpOperation) == path
+
+        where:
+        httpOperation | path
+        "get"         | "/health"
```
Comment (szpak) [inline src/test/groovy/com/ofg/geb/HealthEndpointUISpec.groovy]: I would generate what list in Groovy.

File: src/test/groovy/com/ofg/geb/HealthEndpointUISpec.groovy [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+class HealthEndpointUISpec extends GebSpec {
+
+    def "Setup step"() {
+        when:
+        to SwaggerUIHomePage
+        then:
+        at SwaggerUIHomePage
+        showHealthMVCEndpoints.click()
+        waitFor { heathEndpointsTable.displayed }
+        waitFor { $("#resource_health-mvc-endpoint li.trace span.path a").displayed }
```
Comment (szpak) [inline src/test/groovy/com/ofg/geb/HealthEndpointUISpec.groovy]: Why not put it into a Page object?

File: src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+class AcceptanceSwaggerUISpec extends GebSpec {
+
```
Comment (szpak) [inline src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy]: Would it be useful to fail fast with `@StepWise`?

File: build.gradle (hunk 1 above)
Comment (thiefik) [inline build.gradle]: Good idea, done.

File: src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy (hunk 2 above)
Comment (thiefik) [inline src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy]: Ok, done

File: src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy (hunk 3 above)
Comment (thiefik) [inline src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy]: ok

File: src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy (hunk 4 above)
Comment (thiefik) [inline src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy]: Thanks

File: src/test/groovy/com/ofg/geb/pages/SwaggerUIHomePage.groovy (hunk 5 above)
Comment (thiefik) [inline src/test/groovy/com/ofg/geb/pages/SwaggerUIHomePage.groovy]: Page file has now only '/swagger', rest is in 'baseUrl' property in GebConfig.groovy

File: src/test/groovy/com/ofg/geb/MicroserviceControllerUISpec.groovy (hunk 6 above)
Comment (thiefik) [inline src/test/groovy/com/ofg/geb/MicroserviceControllerUISpec.groovy]: Fixed

File: src/test/groovy/com/ofg/geb/MicroserviceControllerUISpec.groovy (hunk 7 above)
Comment (thiefik) [inline src/test/groovy/com/ofg/geb/MicroserviceControllerUISpec.groovy]: Ok, moved

File: src/test/groovy/com/ofg/geb/MicroserviceControllerUISpec.groovy (hunk 8 above)
Comment (thiefik) [inline src/test/groovy/com/ofg/geb/MicroserviceControllerUISpec.groovy]: Changed to:
Loader.getResource("microservice.json")

File: src/test/groovy/com/ofg/geb/HealthEndpointUISpec.groovy (hunk 9 above)
Comment (thiefik) [inline src/test/groovy/com/ofg/geb/HealthEndpointUISpec.groovy]: This form is easier to read. I modified from '|' to '||'

File: src/test/groovy/com/ofg/geb/HealthEndpointUISpec.groovy (hunk 10 above)
Comment (thiefik) [inline src/test/groovy/com/ofg/geb/HealthEndpointUISpec.groovy]: Done

File: src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy (hunk 11 above)
Comment (thiefik) [inline src/test/groovy/com/ofg/geb/AcceptanceSwaggerUISpec.groovy]: Good idea

**Comment by thiefik** [comment]
Issues from review have been fixed.

**Comment by thiefik** [comment]
Added to master in this commit:
https://github.com/4finance/boot-microservice/commit/e5c49bf01012b33d84a96ce80d20be676c958e88

**Pull request closed without merging**